web/hooks/push.go

```diff
@@ -0,0 +1,259 @@
+	wh.logger.Debugf("Received push event for branch reference: %s (user's default branch: %s)",
+		payload.GetRef(), payload.GetRepo().GetDefaultBranch())
+
+	repos, err := wh.db.GetRepositories(&pb.Repository{RepositoryID: uint64(payload.GetRepo().GetID())})
+	if err != nil {
+		wh.logger.Errorf("Failed to get repository by remote ID %d from database: %v", payload.GetRepo().GetID(), err)
+		return
+	}
+	if len(repos) != 1 {
+		wh.logger.Debugf("Ignoring push event for unknown repository: %s", payload.GetRepo().GetFullName())
+		return
+	}
+	repo := repos[0]
+	wh.logger.Debugf("Received push event for repository %v", repo)
+
+	course, err := wh.db.GetCourseByOrganizationID(repo.OrganizationID)
+	if err != nil {
+		wh.logger.Errorf("Failed to get course from database: %v", err)
+		return
+	}
+	wh.logger.Debugf("For course(%d)=%v", course.GetID(), course.GetName())
+
+	switch {
+	case repo.IsTestsRepo():
+		if !isDefaultBranch(payload) {
+			wh.logger.Debugf("Ignoring push event for non-default branch: %s", payload.GetRef())
+			return
+		}
+		// the push event is for the 'tests' repo, which means that we
+		// should update the course data (assignments) in the database
+		assignments.UpdateFromTestsRepo(wh.logger, wh.db, course)
+
+	case repo.IsUserRepo():
+		wh.logger.Debugf("Processing push event for user repo %s", payload.GetRepo().GetName())
+		if !isDefaultBranch(payload) {
+			wh.logger.Debugf("Ignoring push event for non-default branch: %s", payload.GetRef())
+			return
+		}
+		wh.updateLastActivityDate(repo.UserID, course.ID)
+		assignments := wh.extractAssignments(payload, course)
+		for _, assignment := range assignments {
+			if !assignment.IsGroupLab {
+				// only run non-group assignments
+				wh.runAssignmentTests(assignment, repo, course, payload)
+			} else {
+				wh.logger.Debugf("Ignoring assignment: %s, pushed to user repo: %s", assignment.GetName(), payload.GetRepo().GetName())
+			}
+		}
+
+	case repo.IsGroupRepo():
+		wh.logger.Debugf("Processing push event for group repo %s", payload.GetRepo().GetName())
+		jobOwner, _, err := wh.db.GetUserByCourse(course, payload.GetSender().GetLogin())
+		if err != nil {
+			wh.logger.Errorf("Failed to find user %s in course %s: %v", payload.GetSender().GetLogin(), course.GetName(), err)
+			return
+		}
+		wh.updateLastActivityDate(jobOwner.ID, course.ID)
+		assignments := wh.extractAssignments(payload, course)
+		for _, assignment := range assignments {
+			if assignment.IsGroupLab {
+				// only run group assignments
+				// TODO(Espeland): Need to find a way to have these tests run on the correct branch
+				results := wh.runAssignmentTests(assignment, repo, course, payload)
+
+				if !isDefaultBranch(payload) {
+					// Attempt to find the pull request for the branch, if it exists,
+					// and then assign reviewers to it, if the branch task score is higher than the assignment score limit.
+					wh.handleTaskPush(payload, results, assignment, course, repo)
```

Comment:
What happens if an assignment is graded manually? Results would be nil here.
When do we assign reviewers for a manually graded assignment?
Does manually graded simply mean that a teacher has to approve the assignment for it to pass, 
and that there are no tests? Or is it just manually approved, but with no tests?

Reply:
It seems manually graded implies that an assignment has no tests.
But an assignment that uses pull requests requires tests to work.
An assignment would therefore have to be set as automatically approved, which kind of contradicts the pull request approval process.

Reply:
Apperantly, manually graded simply means that the reviewers field is greater than 0.
Any assignment using pull requests and so on can therefore not be manually graded, since they produce no test results.

Reply:
Indeed. 
Manual grading has teachers / TAs reviewing submitted code, and assigning a grade (passed/failed) to a number of criteria:
![image](https://user-images.githubusercontent.com/1346291/167308912-5b903ee3-51fd-49d0-90d8-1e012de79bc7.png)
These criteria are defined in a `criteria.json` file. Look [here](https://github.com/quickfeed/quickfeed/blob/master/doc/teacher.md#reviewing-student-submissions) for more information if interested.

These submissions will not be tested. The submissions will be assigned a score based on how many of the criteria of a review have been approved.

The above method is only used if the `reviewers` field in `assignment.yml` is greater than 0.

For submissions that are tested, they will be graded (`Approved`, `Revision`, `Failed`) automatically if `autoapprove` is set to `true`. Otherwise, teachers / TAs have to manually assign a grade to the submission.

Reply:
Thank you for further info.

I was really scratching my head the other day, trying to understand the difference between graded and approved.
The specific example where this caused confusion is the following code:
```go
if assignment.GradedManually() {
wh.logger.Debugf("Assignment %s for course %s is manually reviewed", assignment.Name, course.Name)
if _, err := runData.RecordResults(wh.logger, wh.db, nil); err != nil {
	wh.logger.Error(err)
}
return nil
}
```
Intuitively, I imagined thinking that the GradedManually method checked whether an assignment is manually approved.
Would it maybe be better to have this method called ReviewedManually?
Even the log seems to imply this.

Reply:
Clarifying my confusion between manually graded and manually approved.

To me, a manually graded **Assignment** would mean an assignment that gets a grade from A-F.
Therfore, any grade above F would imply that it has passed, and is therby also approved.